File: src/main/kotlin/uz/rms/modules/v1/settings/service/SettingsService.kt
package uz.rms.modules.v1.settings.service

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import uz.rms.modules.v1.settings.dto.*
import uz.rms.modules.v1.settings.model.SettingType
import uz.rms.modules.v1.settings.model.Settings

interface SettingsService {
    // Basic CRUD operations
    fun getAll(): List<SettingsDto>
    fun getById(id: Long): SettingsDto
    fun getByKey(key: String): SettingsDto?
    fun create(settingsDto: SettingsDto): SettingsDto
    fun update(id: Long, settingsDto: SettingsDto): SettingsDto
    fun delete(id: Long)
    
    // Grouped settings
    fun getGroupedSettings(): List<SettingsGroupDto>
    fun getSettingsByType(type: SettingType): List<SettingsDto>
    fun getSettingsByGroup(groupName: String): List<SettingsDto>
    
    // Specific settings
    fun getCompanySettings(): CompanySettingsDto
    fun updateCompanySettings(settings: CompanySettingsDto): CompanySettingsDto
    
    fun getSmtpSettings(): SMTPSettingsDto
    fun updateSmtpSettings(settings: SMTPSettingsDto): SMTPSettingsDto
    
    fun getBackupSettings(): BackupSettingsDto
    fun updateBackupSettings(settings: BackupSettingsDto): BackupSettingsDto
    
    // Search and filter
    fun search(query: String, pageable: Pageable): Page<SettingsDto>
    fun getPublicSettings(): List<SettingsDto>
}

File: frontend/src/services/settingsService.ts
import api from '../lib/api';
import { 
  SettingsDto, 
  CompanySettingsDto, 
  SMTPSettingsDto, 
  BackupSettingsDto,
  SettingsResponse,
  SettingType,
  BackupFrequency
} from '../types/settings';

export const getSettings = async (): Promise<SettingsDto[]> => {
  const response = await api.get('/api/v1/settings');
  return response.data.data;
};

export const getGroupedSettings = async (): Promise<SettingsResponse> => {
  const response = await api.get('/api/v1/settings/grouped');
  return response.data.data;
};

export const getSettingById = async (id: number): Promise<SettingsDto> => {
  const response = await api.get(`/api/v1/settings/${id}`);
  return response.data.data;
};

export const createSetting = async (setting: Partial<SettingsDto>): Promise<SettingsDto> => {
  const response = await api.post('/api/v1/settings', setting);
  return response.data.data;
};

export const updateSetting = async (id: number, setting: Partial<SettingsDto>): Promise<SettingsDto> => {
  const response = await api.put(`/api/v1/settings/${id}`, setting);
  return response.data.data;
};

export const deleteSetting = async (id: number): Promise<void> => {
  await api.delete(`/api/v1/settings/${id}`);
};

// Company Settings
export const getCompanySettings = async (): Promise<CompanySettingsDto> => {
  const response = await api.get('/api/v1/settings/company');
  return response.data.data;
};

export const updateCompanySettings = async (settings: CompanySettingsDto): Promise<CompanySettingsDto> => {
  const response = await api.put('/api/v1/settings/company', settings);
  return response.data.data;
};

// SMTP Settings
export const getSmtpSettings = async (): Promise<SMTPSettingsDto> => {
  const response = await api.get('/api/v1/settings/smtp');
  return response.data.data;
};

export const updateSmtpSettings = async (settings: SMTPSettingsDto): Promise<SMTPSettingsDto> => {
  const response = await api.put('/api/v1/settings/smtp', settings);
  return response.data.data;
};

// Backup Settings
export const getBackupSettings = async (): Promise<BackupSettingsDto> => {
  const response = await api.get('/api/v1/settings/backup');
  return response.data.data;
};

export const updateBackupSettings = async (settings: BackupSettingsDto): Promise<BackupSettingsDto> => {
  const response = await api.put('/api/v1/settings/backup', settings);
  return response.data.data;
};

// Search and filter
export const searchSettings = async (query: string, page: number = 0, size: number = 20) => {
  const response = await api.get('/api/v1/settings/search', {
    params: { query, page, size }
  });
  return response.data.data;
};

export const getSettingsByType = async (type: SettingType): Promise<SettingsDto[]> => {
  const response = await api.get(`/api/v1/settings/type/${type}`);
  return response.data.data;
};

export const getSettingsByGroup = async (groupName: string): Promise<SettingsDto[]> => {
  const response = await api.get(`/api/v1/settings/group/${encodeURIComponent(groupName)}`);
  return response.data.data;
};
